### PR TITLE
custompayloads: Add Initial API (actions)

### PR DESCRIPTION
**Pull request by kingthorin**
## Overview

- CHANGELOG > Added note.
- CustomPayloadsApi > The new API functionality.
- CustomPayloadsApiUnitTest > Add general tests.
- Messages.properties > The descriptor keys/values for the new API components.
- api.html > New API help content.
- index.xml, map.jhm, toc.xml > Link/configure new API help content.

## Checklist
- [X] Update help
- [X] Update changelog
- [X] Run `./gradlew spotlessApply` for code formatting
- [X] Write tests
- [X] Check code coverage
- [X] Sign-off commits
- [X] Squash commits
- [X] Use a descriptive title

### CODE REVIEW

**Comment by kingthorin** [comment]
Addressed IRC discussion.

File: addOns/custompayloads/src/main/javahelp/org/zaproxy/zap/extension/custompayloads/resources/help/contents/api.html [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+	<ul>
+		<li>customPayloads(category): Lists all the payloads currently loaded (category, payload, enabled state). Optionally filtered by category.</li>
+		<li>customPayloadsCategories: Lists all the available categories.</li>
+	</ul>
+
+	<h3>Actions</h3>
+
+	<ul>
+		<li>addCustomPayload (category*, payload*): Add a custom payload (enabled when added).</li>
```
Comment (thc202) [inline addOns/custompayloads/src/main/javahelp/org/zaproxy/zap/extension/custompayloads/resources/help/contents/api.html:21]: `Adds`? (and so on) to match `Lists` above? Another thing to normalize.

File: addOns/custompayloads/src/main/javahelp/org/zaproxy/zap/extension/custompayloads/resources/help/contents/api.html [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+<TITLE>Custom Payloads API</TITLE>
+</HEAD>
+<BODY>
+	<H1>Cusotm Payloads API</H1>
+	The following views are available via the API:
+
+	<h3>Views</h3>
+
+	<ul>
+		<li>customPayloads(category): Lists all the payloads currently loaded (category, payload, enabled state). Optionally filtered by category.</li>
```
Comment (thc202) [inline addOns/custompayloads/src/main/javahelp/org/zaproxy/zap/extension/custompayloads/resources/help/contents/api.html:14]: Make the resource message the same as this? (same for the following endpoint)

File: addOns/custompayloads/src/main/java/org/zaproxy/zap/extension/custompayloads/CustomPayloadsApi.java [hunk 3]
```diff
@@ -86,19 +107,125 @@ public ApiResponse handleApiView(String name, JSONObject params) throws ApiExcep
...
+                        extension.getParam().setPayloads(payloads);
+                        break;
+                    }
+                }
+                return success ? ApiResponseElement.OK : ApiResponseElement.FAIL;
+
+            case ACTION_ADD:
+                category = getValidatedCategory(params.getString(PARAM_CATEGORY));
+                payload = params.getString(PARAM_PAYLOAD);
+                if (payload.isBlank()) {
```
Comment (thc202) [inline addOns/custompayloads/src/main/java/org/zaproxy/zap/extension/custompayloads/CustomPayloadsApi.java:146]: The GUI allows to add it (and IMO it makes sense to allow it).

File: addOns/custompayloads/src/main/java/org/zaproxy/zap/extension/custompayloads/CustomPayloadsApi.java [hunk 4]
```diff
@@ -86,19 +107,125 @@ public ApiResponse handleApiView(String name, JSONObject params) throws ApiExcep
...
+                return ApiResponseElement.OK;
+
+            case ACTION_DISABLE_PAYLOAD:
+                category = getValidatedCategory(params.getString(PARAM_CATEGORY));
+                payload = params.getString(PARAM_PAYLOAD);
+                handlePayload(category, payload, x -> x.setEnabled(false));
+                return ApiResponseElement.OK;
+
+            default:
+                throw new ApiException(ApiException.Type.BAD_VIEW);
```
Comment (thc202) [inline addOns/custompayloads/src/main/java/org/zaproxy/zap/extension/custompayloads/CustomPayloadsApi.java:168]: Needs update.

**Comment by kingthorin** [comment]
Addressed all those, hopefully 😋

File: addOns/custompayloads/src/main/javahelp/org/zaproxy/zap/extension/custompayloads/resources/help/contents/api.html (hunk 1 above)
Comment (thc202) [reply addOns/custompayloads/src/main/javahelp/org/zaproxy/zap/extension/custompayloads/resources/help/contents/api.html]: Update in the resource messages too?

File: addOns/custompayloads/src/main/java/org/zaproxy/zap/extension/custompayloads/CustomPayloadsApi.java (hunk 3 above)
Comment (thc202) [reply addOns/custompayloads/src/main/java/org/zaproxy/zap/extension/custompayloads/CustomPayloadsApi.java]: Note that the `payload` will have to be made optional as empty is considered missing if required.

**Comment by kingthorin** [comment]
Tweaked.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Review by psiinon — APPROVED** [review]

**Pull request merged**